clang/test/SemaCXX/builtin-is-bitwise-cloneable.cpp

```diff
@@ -0,0 +1,41 @@
+// RUN: %clang_cc1 -std=c++20 -fsyntax-only -verify %s
+
+// Scalar types are bitwise clonable.
+static_assert(__is_bitwise_cloneable(int));
+static_assert(__is_bitwise_cloneable(int*));
+// array
+static_assert(__is_bitwise_cloneable(int[10]));
+
+// non-scalar types.
+static_assert(!__is_bitwise_cloneable(int&));
+
+
+struct Forward; // expected-note 2{{forward declaration of 'Forward'}}
+static_assert(!__is_bitwise_cloneable(Forward)); // expected-error {{incomplete type 'Forward' used in type trait expression}}
+
+struct Foo { int a; };
+static_assert(__is_bitwise_cloneable(Foo));
+
+struct DynamicClass { virtual int Foo(); };
+static_assert(__is_bitwise_cloneable(DynamicClass));
+
+struct Bar { int& b; }; // trivially copyable
+static_assert(__is_trivially_copyable(Bar));
+static_assert(__is_bitwise_cloneable(Bar));
```

Comment:
I don't quite get the rule here.
If reference members make the class non-bitwise-cloneable, why isn't that assertion firing?
If they don't, why is the next assertion a problem (the two classes only differ by presence of constructor, which we choose to ignore in bitwise-cloneability).

Reply:
I think it makes a lot of sense to make this a strict superset of trivially copyable types.
But why isn't next example bitwise-cloneable then?

```cpp
struct Bar2 { Bar2(const Bar2&); int& b; };
static_assert(!__is_trivially_copyable(Bar2));
static_assert(!__is_bitwise_cloneable(Bar2)); 
```
`Bar2` is not trivially copyable because of the presence of a user-defined constructor, we chose to drop that requirement for the new trait. Why isn't the type bitwise-cloneable?

Reply:
Because of the reference member `int & b`.

Reference types are currently excluded in bitwise-cloneable types (references are just alias, and they cannot be reset once initialized).

Reply:
But the reference member in `Bar` does not stop it from being bitwise-cloneable.
We should make up our mind about it: do reference members stop the type from being bitwise-cloneable or not?